pinot-core/src/main/java/org/apache/pinot/core/segment/processing/mapper/SegmentMapper.java

```diff
@@ -141,28 +146,43 @@ private Map<String, GenericRowFileManager> doMap()
               RecordReaderFactory.getRecordReader(recordReaderFileConfig._fileFormat, recordReaderFileConfig._dataFile,
                   recordReaderFileConfig._fieldsToRead, recordReaderFileConfig._recordReaderConfig);
           mapAndTransformRow(recordReader, reuse, observer, count, totalCount);
+          _recordReaderFileConfigs.get(i)._recordReader = recordReader;
         } finally {
-          if (recordReader != null) {
+          if (recordReader != null && !recordReader.hasNext()) {
```

Comment:
Let's add the comment on why we don't close the reader if recordReader has more records to read.

Reply:
decoupled the details of closing the recordreaders in `RecordReaderFileConfig` following Manish's suggestion.